eventstreamrpc/source/EventStreamClient.cpp

```diff
@@ -0,0 +1,1226 @@
+            : m_responseType(ERROR_RESPONSE)
+        {
+            m_responseResult.m_error = std::move(error);
+        }
+
+        TaggedResponse::TaggedResponse(TaggedResponse &&taggedResponse) noexcept
+        {
+            if (m_responseType == EXPECTED_RESPONSE)
+            {
+                m_responseResult.m_response = std::move(taggedResponse.m_responseResult.m_response);
+            }
+            else if (m_responseType == ERROR_RESPONSE)
+            {
+                m_responseResult.m_error = std::move(taggedResponse.m_responseResult.m_error);
+            }
+        }
+
+        TaggedResponse::operator bool() const noexcept
+        {
+            if (m_responseType == EXPECTED_RESPONSE)
+            {
+                return true;
+            }
+            else if (m_responseType == ERROR_RESPONSE)
+            {
+                return false;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        OperationResponse *TaggedResponse::GetResponse()
+        {
+            if (m_responseType == EXPECTED_RESPONSE)
+            {
+                return m_responseResult.m_response.get();
+            }
+            else
+            {
+                return nullptr;
+            }
+        }
+
+        OperationError *TaggedResponse::GetError()
+        {
+            if (m_responseType == ERROR_RESPONSE)
+            {
+                return m_responseResult.m_error.get();
+            }
+            else
+            {
+                return nullptr;
+            }
+        }
+
+        std::future<TaggedResponse> ClientOperation::GetResponse() noexcept
+        {
+            return m_initialResponsePromise.GetFuture();
+        }
+
+        const EventStreamHeader *ClientOperation::GetHeaderByName(
+            const Crt::List<EventStreamHeader> &headers,
+            const Crt::String &name) noexcept
+        {
+            for (auto it = headers.begin(); it != headers.end(); ++it)
+            {
+                if (name == it->GetHeaderName())
+                {
+                    return &(*it);
+                }
+            }
+            return nullptr;
+        }
+
+        int ClientOperation::HandleData(const Crt::String &modelName, const Crt::Optional<Crt::ByteBuf> &payload)
+        {
+            ExpectedResponseFactory responseFactory = nullptr;
+
+            /* Responses after the first message don't necessarily have the same shape as the first. */
+            if (m_messageCount == 1)
+            {
+                responseFactory = m_responseRetriever.GetLoneResponseFromModelName(GetModelName());
+            }
+            else
+            {
+                responseFactory = m_responseRetriever.GetStreamingResponseFromModelName(GetModelName());
+            }
+
+            if (responseFactory == nullptr)
+            {
+                return EVENT_STREAM_RPC_UNMAPPED_DATA;
+            }
```

Comment:
The control flow for handling incoming data should defer to the Smithy model (through generated code+interfaces). This flow assumes that if it got a second message it must be a streaming operation, then if it can't find a handler for it, it it received unmapped data. This is confusing and should be avoided.